test/bpf_tests/bpf_test.go

```diff
@@ -0,0 +1,470 @@
+					return
+				}
+
+				t.Log(dm.Message(linkCache))
+			}
+		}()
+	}
+
+	for name, progs := range testNameToPrograms {
+		t.Run(name, subTest(progs, coll.Maps[suiteResultMap]))
+	}
+
+	if globalLogReader != nil {
+		// Give the global log buf some time to empty
+		// TODO: replace with flush on the buffer, as soon as cilium/ebpf supports that
+		time.Sleep(50 * time.Millisecond)
+	}
+}
+
+func loadAndPrepSpec(t *testing.T, elfPath string) *ebpf.CollectionSpec {
+	spec, err := ebpf.LoadCollectionSpec(elfPath)
+	if err != nil {
+		t.Fatal("load spec: ", elfPath, err)
+	}
+
+	for _, mspec := range spec.Maps {
+		// Drain extra info from map specs, cilium/ebpf will complain if we don't
+		if mspec.Extra != nil {
+			io.ReadAll(mspec.Extra)
+		}
+
+		// Remove BTF from maps as a workaround since some maps, like 'cilium_xdp_scratch' will return a
+		// "map create: invalid argument" error if we don't
+		mspec.BTF = nil
+	}
+
+	// Detect program type mismatches
+	var progTestType ebpf.ProgramType
+	for _, spec := range spec.Programs {
+		if progTestType != spec.Type {
+			if progTestType == ebpf.UnspecifiedProgram {
+				progTestType = spec.Type
+				continue
+			}
+			if spec.Type == ebpf.UnspecifiedProgram {
+				continue
+			}
+
+			t.Fatalf(
+				"File '%s' contains both '%s' and '%s' program types, "+
+					"only one program type per ELF file allowed:",
+				elfPath,
+				progTestType,
+				spec.Type,
+			)
+		}
+	}
+
+	if progTestType == ebpf.UnspecifiedProgram {
+		t.Fatalf("File '%s' only contains unspecified program types", elfPath)
+	}
+
+	// Give all tail call programs the same program type as the test programs
+	for _, spec := range spec.Programs {
+		spec.Type = progTestType
+	}
+
+	return spec
+}
+
+// Fill the tail calls map with the tail calls based on the calls .id and names of the program sections.
+func loadCallsMap(spec *ebpf.CollectionSpec, coll *ebpf.Collection) error {
+	callMap, found := coll.Maps[callsMapName]
+	if !found {
+		// If we can't find the map, tailcalls aren't required for the current tests
+		return nil
+	}
+
+	for name, prog := range coll.Programs {
+		if strings.HasPrefix(spec.Programs[name].SectionName, callsMapID) {
+			indexStr := strings.TrimPrefix(spec.Programs[name].SectionName, callsMapID)
+			index, err := strconv.Atoi(indexStr)
+			if err != nil {
+				return fmt.Errorf("atoi tail call index: %w", err)
+			}
+
+			index32 := uint32(index)
+			err = callMap.Update(&index32, prog, ebpf.UpdateAny)
+			if err != nil {
+				return fmt.Errorf("update tailcall map: %w", err)
+			}
+		}
+	}
+
+	return nil
+}
+
+type programSet struct {
+	setupProg *ebpf.Program
+	checkProg *ebpf.Program
+}
+
+var checkProgRegex = regexp.MustCompile(`[^/]+/test/([^/]+)/((?:check)|(?:setup))`)
+
+const (
+	ResultSuccess = 1
+
+	suiteResultMap = "suite_result_map"
+
+	callsMapName = "test_cilium_calls_65535"
+	// TODO we should read the .id field from the maps BTF, but the current cilium/ebpf version doesn't make the raw
```

Comment:
The .id field is not emitted into BTF, it's in the first 4 bytes of MapSpec.Extra. Again, https://github.com/cilium/cilium/pull/19159/commits/7aaec16d6076e6f6ac562bfc4bdaf2650a42733d takes care of that.